### ISSUE TITLE
transformer output nan in last few layer

### ISSUE THREAD

**Issue by 0rchard** [zhihu/cuBERT#71]
pytorch model 转成 tf model之后, 使用cuBert加载进行pooler_output运算,
发现从layer_8 以后的logits基本都是nan.

能提供一些线索吗?

**Comment by 0rchard** [comment]
i found nan is generate by softmax, due to numerical stability problem.

i replace mkl softmax using log sum exp trick, problem solved.

`    
   template<>
    void softmax_<float>(float *in,
                         float *out,
                         const int batch_size,
                         const int channel,
                         float *sum_gpu,
                         void *stream) {
        for (int batch_idx = 0; batch_idx < batch_size; ++batch_idx) {
            float avg = 0;
            for (int i = batch_idx * channel; i < (batch_idx + 1) * channel; ++i) {
                avg += in[i];
            }
            avg = avg / channel;
            float log_sum = 0;
            for (int i = batch_idx * channel; i < (batch_idx + 1) * channel; ++i) {
                log_sum += expf(in[i] - avg);
            }
            log_sum = logf(log_sum);
            for (int i = batch_idx * channel; i < (batch_idx + 1) * channel; ++i) {
                out[i] = expf(out[i] - avg - log_sum);
            }
        }
    }
`

ref: 
https://stackoverflow.com/questions/9906136/implementation-of-a-softmax-activation-function-for-neural-networks
